### ISSUE TITLE
List of potential changes to Blaze

### ISSUE THREAD

**Issue by mitar** [meteor/blaze#75]
Just a list of some ideas I have what we could change in Blaze to make it a bit better, keep the good parts and move on. Some of them might be breakable so we should see how to integrate this. Personally, I would like to see that we do just incremental changes. One thing to do that would be to introduce components. Then `<template>`s could be seen simply as a HTML fragments with placeholders you can reuse, without any logic. This is what I also started with Blaze Components.

But the new components (in contrast with Blaze Components) would have some changes. I will be making it a lit of comments for each of those so that community can upvote/downvote them.

Please write only ideas (and variations of existing ideas). If you want to discuss details of an idea implementation, process, roadmap, open a ticket for that particular idea (we can update then its comment here with a link to the ticket).

What are your ideas? What would you like to improve in Blaze?

**Comment by mitar** [comment]
You would have to explicitly declare which arguments a component is receiving (as a dict).

**Comment by mitar** [comment]
No implicit passing of a data context.

**Comment by stubailo** [comment]
I'd add inline event binding like react has, and get rid of the event map style.

**Comment by mitar** [comment]
The values of the arguments would be reactive, but not the whole data context (so we get immediately one level deeper equality comparison, instead of top-level data context which changes always).

**Comment by mitar** [comment]
You bind events inside a template (like Blaze Components), not using event maps (they should still be supported for trickier things).

**Comment by mitar** [comment]
We could offer easy way to cache results of reruns of reactivity and if it is the same, we do not have to continue.

**Comment by mitar** [comment]
We could be doing diff between new and old version of internal HTMLJS and then apply just changes (like virtual DOM), maybe this is already done, not sure.

**Comment by mitar** [comment]
Or use something like incremental DOM.

**Comment by mitar** [comment]
For that to work, we introduce a "pure" component which does not allow any mingling with jQuery inside (no $ selector available) and all changes are result of just result of helpers and stuff.

**Comment by mitar** [comment]
This is allows most things to do, the only reason I have to mingle with DOM is animations, so we should provide some way to do that.

**Comment by mitar** [comment]
Arguments to methods can still be list of arguments of keyword arguments, but keyword arguments get wrapped into a normal JavaScript object which is an instance of Spacebars.kw or something, so that it looks like `options` object or something, you do not have to care or you do about the fact that it is an instance of that prototype.

**Comment by mitar** [comment]
You get this object only when somebody really passed keyword arguments to your helper.

**Comment by mitar** [comment]
We deprecate `#with` and simple `#each`, but we simplify that all this is simply part of current scope, no fancy things of data contexts, lexical scopes, and so on, things would simple resolve just: is it in current scope, or is it a method on current component instance.

**Comment by mitar** [comment]
Components try to persist as long as possible, only their rendered HTML is being changed, only their arguments change in values (so you can keep state in them).

**Comment by stubailo** [comment]
Crazy idea: Add a way to write template code with a script tag inside the template itself.

**Comment by mitar** [comment]
When removing components we go from outside in, so that we might remove everything first, and not that we first remove inner elements, and then outside, and so on which takes a long time on route switching.

**Comment by mitar** [comment]
We introduce normal method calling inside helpers, so `{{foo something(a) foo=bar}}` (we can discuss exact syntax).

**Comment by mitar** [comment]
> I'd add inline event binding like react has, and get rid of the event map style.

This is this:

> You bind events inside a template (like Blaze Components), not using event maps (they should still be supported for trickier things).

It is already in Blaze Components. No? Or would you want something else?

**Comment by mitar** [comment]
> Crazy idea: Add a way to write template code with a script tag inside the template itself.

I was thinking that we could have JS code and then simply top-level `<template>` inside same file. We would still have code and template split, but it could be in the same file. But yes, we could wrap the code into `<script>` to be more compatible with various IDEs.

**Comment by caioreis** [comment]
@mitar are you thinking in someway merge Blaze Components here? Or the ideia is to start implementing this stuff from scratch?

**Comment by mitar** [comment]
> @mitar are you thinking in someway merge Blaze Components here? Or the ideia is to start implementing this stuff from scratch?

I do not know yet how we will implement this, but let's keep this issue on topic of ideas. So please write ideas. If you want to discuss details of an idea open a ticket for that particular idea (I can update then its description with a link to the ticket).

**Comment by stubailo** [comment]
> It is already in Blaze Components. No? Or would you want something else?

Sure, I meant add it to core. There are some things I don't personally like in BC, so I wouldn't want to use it just for this feature.

**Comment by mitar** [comment]
> Sure, I meant add it to core. There are some things I don't personally like in BC, so I wouldn't want to use it just for this feature.

No, I mean, you want the thing which is also in Blaze Components, or something else? Because I also added on the list that those inline events (like in Blaze Components) should be go to core. But I am unsure if there is some extra detail of event handling which is not covered there.

**Comment by stubailo** [comment]
Yeah, I think it's about the same. I added my comment first though :P

**Comment by mitar** [comment]
Or should we keep template/logic be 1:1. While in Blaze Components they are split (easier to make it backwards compatible with Blaze, and easier to reuse same HTML with different logic), I must say I have never really reused the same template for multiple components. Keeping template/logic 1:1 is then easier because it is clearer what `{{> Foo}}` does.

So, should we keep template/logic 1:1?

**Comment by stubailo** [comment]
I think the right way to reuse templates should be like React does it - you make one template that has the logic, and then it can accept a logicless template as a parameter.

**Comment by mitar** [comment]
Importing modules. It would be great that `{{> foo.Bar}}` loads `foo` module and uses Bar symbol from it. Maybe we could allow relative imports as well.

The downside is that then Meteor imports would be `{{> meteor.foo.Bar}`. And that for local imports you would need to use relative imports always. But this could be as simple as `{{> .foo.Bar}}`.

**Comment by mitar** [comment]
> I think the right way to reuse templates should be like React does it - you make one template that has the logic, and then it can accept a logicless template as a parameter.

So in our case you could also pass a template with logic, but it would just reuse the non-logic part?

**Comment by mitar** [comment]
For imports mentioned above. Should we then instead of `{{> foo.Bar}}` or `{{> foo/Bar}}`?

**Comment by mitar** [comment]
Or `{{> 'foo/Bar'}}` (because `{{> foo}}` would allow import what is in variable `foo`.

**Comment by stubailo** [comment]
> So in our case you could also pass a template with logic, but it would just reuse the non-logic part?

No, I don't think we should ever separate a template from its logic.

> For imports mentioned above. Should we then instead of {{> foo.Bar}} or {{> foo/Bar}}?

Maybe `{{> Bar from './module'}}` or something? I feel like sticking to import syntax from ES2015 is a good idea.

**Comment by mitar** [comment]
> Maybe `{{> Bar from './module'}}` or something? I feel like sticking to import syntax from ES2015 is a good idea.

How you pass arguments to it? :-)

**Comment by stubailo** [comment]
"or something" 😛 

I also feel like it's a good idea to put the imports at the top rather than inline.

**Comment by mitar** [comment]
Crazy idea then:

```handlebars
<script>
  import Bar from './module'
</script>

<template name="fooTemplate">
  {{> Bar}}
  {{hello}}
</template>

class Foo extends fooTemplate {
  hello() {
    return "awesome";
  }
}
```

**Comment by mitar** [comment]
And arguments:

```handlebars
<template name="fooTemplate">
  {{> Bar data=bar parent=this}}
</template>
```
In `Bar`:

```javascript
class Bar extends barTemplate {
  hello() {
    return this.parent().giveMeSomething() + ": " + this.data();
  }
}
```

**Comment by mitar** [comment]
You could also do:

```handlebars
<template name="fooTemplate">
  {{#let Bar=require('./module').Bar}}
    {{> Bar}}
  {{/let}}
  {{hello}}
</template>
```

And:

```handlebars
<template name="fooTemplate">
  {{> require('./module').Bar}}
  {{hello}}
</template>
```

**Comment by mitar** [comment]
So maybe a simple way to do is that `{{> symbol}}` requires that symbol is a variable. It cannot really be a string anymore. Simply `"symbol"` is not a Blaze template so inclusion fails.

Now, that symbol you can import inside a template, or in the same module (JavaScript code in the same file is part of the same module).

So:

```
<script>
  import Bar from './module'
</script>

<template name="fooTemplate">
  {{> Bar}}
  {{hello}}
</template>

<script>
  class Foo extends fooTemplate {
    hello() {
      return "awesome";
    }
  }
</script>
```

Is simply compiled into:

```javascript

import Bar from './module';

const fooTemplate = spacebarsCompiledTemplate();

class Foo extends fooTemplate {
  hello() {
    return "awesome";
  }
}
```

And for backwards compatibility we can make so that you can somehow easily enable (or disable?) that all templates defined are available as variables in a global scope of all templates. This could even be a global helper which returns that template. So instead of `{{> fooBar}}` resolving `fooBar` in a special way as a named template, it would simply search a scope for `fooBar` and global scope would contain a symbol, a reference to a template, with that name.

**Comment by mitar** [comment]
Another backwards compatible way could be have argument to a template named `export`, if is `true`, it exports a template and do not make it global, if is false it does not export a template, and if it is missing it is a global template (backwards compatibility).

```handlebars
<template name="fooTemplate" export="true">
  {{> Bar}}
  {{hello}}
</template>
```

**Comment by mitar** [comment]
Or even simpler, we introduce `<component>`, and components get `export` argument if you want to export them, otherwise they are local. (Or it could be that they are exported by default.)

And `<component>` could have all other backwards incompatible changes as well.

**Comment by MiroHibler** [comment]
Guys, wouldn't a discussion like this be more appropriate on Slack (or forum or hackpad), so we can attract more broader audience, not just the contributors (I'd guess not all Blaze users are also contributors to the code)?

I know @mitar feels very much at home here ;) but until we reach the execution point, with concrete tasks and issues, I think we should do such discussions somewhere else...

BTW, how relevant is [this](https://meteor.hackpad.com/Proposal-for-Blaze-2-bRAxvfDzCVv) hackpad these days?

**Comment by maxenceC** [comment]
>We could be doing diff between new and old version of internal HTMLJS and then apply just changes (like virtual DOM), maybe this is already done, not sure.

@mitar, Isn't this what ember did to their handlebars templates with the glimmer engine ? : https://github.com/tildeio/glimmer

**Comment by alexandesigner** [comment]
> BTW, how relevant is this hackpad these days?

I think there's a good discussion and a material that maybe we can explore.

>We could adapt it to blaze.

@maxenceC, I like the idea... a while ago I suggested the [Incremental DOM](https://google.github.io/incremental-dom/) by knowing a little more and be relatively simpler, but maybe we could take a look at the [Glimmer](https://github.com/tildeio/glimmer) even the similarity with the Blaze who uses Spacebars.

**Comment by mitar** [comment]
> Guys, wouldn't a discussion like this be more appropriate on Slack (or forum or hackpad)

I do not think so, because ideas tend to get lost there. And people cannot nicely upvote and downvote. I agree that Slack is great for ironing out bugs in real-time (like helping with why tests are failing), but one my main issue with MDG's development process was that things were designed and decided in closed sessions. And Slack to me is closed: if you are not online at the time a discussion is happening, it is more or less history and you will not find it later on.

GitHub is also great because even if people come around in a month they can still vote on things. And many of ideas here will be probably done in a year or two, so they have to stick around for a long time.

And hackpads again split attention of the community. They are great when you want to get some document out. Maybe useful when we will be working on a roadmap and concrete syntax/changes discussion.

> Isn't this what ember did to their handlebars templates with the glimmer engine

It seems so.

> We could adapt it to blaze.

It would be probably more work than worth. We already have most of the pieces available.

**Comment by mitar** [comment]
(But let's keep implementation details of ideas out of this ticket. So discussion on how we would implement diffing: using some library, doing it ourselves, and so on.)

**Comment by tcastelli** [comment]
Not sure if this is somewhere or can be achieved without the use of a global JSON.parse helper, but having simple objects as parameters would be great so no need for extra nasty code. 

`eg:  {{> tmp1 param1="hi" param2='{"a":8, "b":3}' }} `

**Comment by mitar** [comment]
I do not see much difference from `{{> tmp1 param1="hi" param2=(a=8 b=3) }}`? Even less characters?

**Comment by tcastelli** [comment]
Sorry, didn't catch that issue/proposal. TBH the syntax is less important to me, so as long as the feature is in I'm fine. If I could choose, maybe something less confusing tham ( would be fine, since "( " is already used for (sub)expressions and mixing both things could be troublesome:

`{{> tmp1 param1="hi" param2=(helper (a=8 b=3)) (ahelper (b="pa")) }}`

**Comment by mitar** [comment]
Let's have the discussion of details in its own issue. This one is mostly to have a list of all possible improvements and upvoting and downvoting.

**Comment by LaughingBubba** [comment]
If the long term view is to be a stand alone reactive view layer, independent of meteor, shouldn't the focus be on removing dependency or forking on minimongo & tracker? 

Or is the view that blaze will always be wedded to meteor "classic"?

Just curious

**Comment by mitar** [comment]
No, moving to other system is definitely in plan, but we can still develop new features in parallel.

See #11 for NPM transition ticket, and #13 for the dependency, we first have to move Tracker and others to NPM. I think we need some champions of pushing those changed forward. And they also need a bit of help from MDG. But let's discuss there.

**Comment by LaughingBubba** [comment]
Thanks so the short answer is not any time soon to impact what I had in mind. 

That's Ok, the reason is my best contribution is likely to be the voice of the n00b. So my plan was draft up a visual/schematic of how blaze works (with meteor) at a high level and then an anatomy of a simple blaze app. I know a lot of this is covered in the blazejs.org site but that's all is the written word. This is about keeping the learning curve as flat as possible.

Is that OK?

**Comment by mitar** [comment]
You should definitely check out [guide](http://guide.meteor.com/) because this is probably what is most suitable for there, no? But I would love to have that for blazejs as well, it seems that blazejs will be then more like "meteor classic" or what. :-) But yea, open a ticket/pull request and let's discuss it there.

**Comment by LaughingBubba** [comment]
The guide is awsome and did not happen overnight. I'm just trying to capture that "wow" moment back in the 1.0 days. You did not need a monster guide back then cuz blaze templates and handlers were quite understandable. Anyways I think these up in rough form and throw it out there and if I've done the job right it'll be self evident. If so then we figure where it sits. Cheers.

**Comment by mitar** [comment]
Sounds amazing! Go for it! :-)

**Comment by alexandesigner** [comment]
Guys.. I don't know if it's the moment, but maybe write a blog (medium)? where would have daily or weekly postings. This could give blaze us a little more visibility and also make new developers interested learn blaze and other things related.. What do you think?

**Comment by mitar** [comment]
>  don't know if it's the moment, but maybe write a blog (medium)? where would have daily or weekly postings. This could give blaze us a little more visibility and also make new developers interested learn blaze and other things related.. What do you think?

I personally do not know much what we would write, but if anyone wants to take a lead on that, it could be great! :-) We could probably create Medium publication and then people could just submit posts they would like to be published under that banner? (Maybe the only requirement could be that guest posts have to be true, about Blaze, a good practice, and not too promotional for your own stuff.) I am not sure what we would already post, but maybe even just simple things, maybe just things from the changelog, an example of a change and how this improves something, maybe shoutouts to projects using Blaze, things like that? So yea, if anyone has ideas for this, it could be interesting. I am trying to focus my limited time on more technical things.

We can open another issue to discuss this in more detail (what would be the rules for publishing there guest posts and things like that).

**Comment by nicooprat** [comment]
If you make the one-file-component happen, please don't forget CSS :) The way VueJS does it looks perfect to me!

PS. Very happy to see Blaze resurrect :D

**Comment by mitar** [comment]
@nicooprat: Can you explain a bit more what VueJS does, or provide a link? I do not know of anything special it is doing about CSS?

**Comment by alexandesigner** [comment]
> We can open another issue to discuss this in more detail (what would be the rules for publishing there guest posts and things like that).

Exactly ... from the most basic level to advanced as well as projects using blaze..

I will create a page on @medium and open an issue to discuss how we could make the publications.

> Can you explain a bit more what VueJS does, or provide a link? I do not know of anything special it is doing about CSS?

I think he is suggesting something as quoted in another issue ... [Single File Components] (https://vuejs.org/guide/application.html#Single-File-Components)

**Comment by mitar** [comment]
> I will create a page on @medium

You mean a publication?

**Comment by mitar** [comment]
> I think he is suggesting something as quoted in another issue ... [Single File Components] (https://vuejs.org/guide/application.html#Single-File-Components)

Oh, yes. `<style>` tag along with `<template>` tag and `<script>` tag. For sure! :-)

We can also think about ways to do `<style>` tag scoped to the component.

**Comment by alexandesigner** [comment]
> You mean a publication?

Exactly

**Comment by mitar** [comment]
On the other hand, it would be even crazier if `<style>` would get the same context as `<template>`. So you could do:

```
<style>
  .button {
    width: {{width}};
  }
</style>
```

**Comment by aadamsx** [comment]
What do you say about using something like these to help manage this Blaze project?  
https://github.com/integrations/feature/project-management

**Comment by mitar** [comment]
> What do you say about using something like these to help manage this Blaze project?

I looked into GitHub projects when they came out, I do not yet see usefulness for that. And for integrations with others I tried them in the past but then none really provided so much benefit to me that it would balance the minus you get: people have to search for things in to places, you divide the focus into two places, beginners have issues knowing where something is, you get sometimes things desynced. Just remember Meteor Trello roadmap as an example. This is why I prefer everything to be in one place. So using issues for both discussions, bug reports, and feature requests. And then using extensively labels and milestones for content-based categorization and time-based categorization, respectively.

But if you have a concrete proposal what you think would improve management, please open a new issue with a suggestion.

I think that for now community is small enough that what we have is manageable. And moreover, any extra tool requires extra resources to keep it useful, and those resources I would prefer to direct at Blaze itself.

**Comment by mitar** [comment]
But you are right, I think it is time to open separate issues for some of the ideas upvoted above. And start discussing details.

**Comment by alexandesigner** [comment]
> What do you say about using something like these to help manage this Blaze project?

I like the idea.. I think some of these tools could improve the visualization and understanding of what should be done and what has been done.. They are integrated with github

**Comment by nicooprat** [comment]
@mitar Yes, I mean style tag in the same file, and the possibility to have that CSS scoped to its component, by generating a random ID (http://vue-loader.vuejs.org/en/features/scoped-css.html). It's using webpack though.

**Comment by mitar** [comment]
I opened tickets for some of the ideas above.

**Comment by aadamsx** [comment]
> I looked into GitHub projects when they came out, I do not yet see usefulness for that (mostly, upvotes are not visible).

I'm using GitHub projects for internal projects and it helps organize the issues.  Basically can can categorize and track issues with GitHub project with a Kanban type board.  I think just having a the following columns will work: 

```> Backlog```

```> Planning```

```> To Do```

```> In Progress```

```> In Review```

```> To Pull```

**Comment by mitar** [comment]
I do not understand what is the advantage over [milestones](https://github.com/meteor/blaze/milestones). It is clear what is slatted for future, and what is the next version (2.2). You have [a nice list as well](https://github.com/meteor/blaze/milestone/4), and if users are assigned, you can see that as well.

> Besides -- it takes nothing away, only adds to our understanding.

Hm, you have to maintain both the roadmap and projects, no? What if they conflict? So what is the advantage over a project over the roadmap?

BTW, this should go to another issue for discussion. I am not completely opposed to using projects, or whatever, but I have not yet been convinced. Mostly because our current mode of operation is mostly that we respond to random pull requests from people. We do not really have a steady development where we would have dedicated people working on stuff. But if you want to know what would be great to work on, you can, or check those lists:

* https://github.com/meteor/blaze/milestone/4
* https://github.com/meteor/blaze/labels/contributions%20welcome
* https://github.com/meteor/blaze/labels/help%20needed

Volunteers welcome! Pull requests as well.

Or what other problem we want to solve with projects? So planing - for me roadmap works. Knowing what to work on - labels and milestone works as well.

**Comment by aadamsx** [comment]
> So what is the advantage over a project over the roadmap?

I thought **Milestones** where to group issues for PRs not for organization and planning?

The new Github **Project** feature is for project management:

![screen shot 2016-09-29 at 12 18 59 pm](https://cloud.githubusercontent.com/assets/5341158/18964550/fb937e18-863e-11e6-93e1-600322d75eb1.png)

> Hm, you have to maintain both the roadmap and projects, no? What if they conflict? 

Technically, nothing will "happen".  I mean, the core of github (besides the repo) are issues.  The issues are used in Milestones and in Projects.  You could have a Project that spans multiple Milestones if you set it up like that.

For example, this is one way to set the Blaze project up so that it spans multiple versions (as we pass up versions, they could be dropped off the board):

![screen shot 2016-09-29 at 12 25 45 pm](https://cloud.githubusercontent.com/assets/5341158/18964788/0416844e-8640-11e6-89c6-09b721f3f5de.png)

**Comment by tcastelli** [comment]
@aadamsx as commented on #127 I think that this could be more automated with webhooks and automatic triggers, so besides "setting the proper labels" nothing more would have to be done(no drag and drop to colums since each label goes to a column to trello or github projects).
Btw, I think we can continue on the other issue so we keep this one for ideas only.

**Comment by aadamsx** [comment]
There -- deleted that for-ya.  

Just having a list of issues is not a high enough abstraction to get a feel for the project for some.  And please don't knock us for wanting more than the minimum. ;)

But let's do it over on #127 shall we?

**Comment by filipenevola** [comment]
I'm closing this issue because it's too old.

If you think this issue is still relevant please open a new one.

Why? We need to pay attention to all the open items so we should keep opened only items where people are involved.
